backend/infrahub/core/constraint/node/runner.py

```diff
@@ -23,19 +23,37 @@ def __init__(
         self.uniqueness_constraint = uniqueness_constraint
         self.relationship_manager_constraints = relationship_manager_constraints
 
+    async def _check(
+        self,
+        db: InfrahubDatabase,
+        node: Node,
+        field_filters: list[str] | None = None,
+        skip_uniqueness_check: bool = False,
+    ) -> None:
+        await node.resolve_relationships(db=db)
+
+        if not skip_uniqueness_check:
+            self.uniqueness_constraint.db = db
+            await self.uniqueness_constraint.check(node, filters=field_filters)
+
+        for relationship_name in node.get_schema().relationship_names:
+            if field_filters and relationship_name not in field_filters:
+                continue
+            relationship_manager: RelationshipManager = getattr(node, relationship_name)
+            await relationship_manager.fetch_relationship_ids(db=db, force_refresh=True)
+            for relationship_constraint in self.relationship_manager_constraints:
+                relationship_constraint.db = db
+                await relationship_constraint.check(relm=relationship_manager, node_schema=node.get_schema())
+
     async def check(
         self, node: Node, field_filters: list[str] | None = None, skip_uniqueness_check: bool = False
     ) -> None:
-        async with self.db.start_session() as db:
-            await node.resolve_relationships(db=db)
-
-            if not skip_uniqueness_check:
-                await self.uniqueness_constraint.check(node, filters=field_filters)
-
-            for relationship_name in node.get_schema().relationship_names:
-                if field_filters and relationship_name not in field_filters:
-                    continue
-                relationship_manager: RelationshipManager = getattr(node, relationship_name)
-                await relationship_manager.fetch_relationship_ids(db=db, force_refresh=True)
-                for relationship_constraint in self.relationship_manager_constraints:
-                    await relationship_constraint.check(relm=relationship_manager, node_schema=node.get_schema())
+        if not self.db.is_transaction:
```

Comment:
We need to be careful here, as far as I remember one of the reason we are starting a new session is that otherwise we can't use the parallel runtime to check the uniqueness constraint and it will impact performance
I could be wrong here, I need to look deeper in the code